generate-serializers.py should support nested conditions

aprotyas opened the pull request:
#### 4ad01b4b62e7120bfc5bf78f62b5946884e2743a
<pre>
generate-serializers.py should support nested conditions
<a href="https://bugs.webkit.org/show_bug.cgi?id=268519">https://bugs.webkit.org/show_bug.cgi?id=268519</a>
<a href="https://rdar.apple.com/122056394">rdar://122056394</a>

Reviewed by Brady Eidson and Alex Christensen.

generate-serializers.py does not currently support nested conditions,
not because of a lack of intention, but just because there hasn&apos;t been a
need. However, with the proposed changes for <a href="https://webkit.org/b/267428">https://webkit.org/b/267428</a>,
the watchOS/tvOS/GTK ports fail to compile because several Apple Pay
enablement conditions are being stacked when declaring data types in
WebCoreArgumentCodersCocoa.serialization.in.

This patch addresses the issue by introducing support for nested
conditions in the serializer generation script. It does so by imbibing
the parse_serialized_types method with the concept of a stack of
conditions, pushing to the stack (or negating the top entry) when an IF
or ELSE is seen, respectively, and only popping from the stack when an
ENDIF is seen. The script now keeps an aggregated condition expression
by walking through the condition stack and logically concatenating
conditions.

* Source/WebKit/Scripts/generate-serializers.py:
(ConditionStackEntry):
(ConditionStackEntry.__init__):
(ConditionStackEntry.expression):
(generate_condition_expression):
(parse_serialized_types):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;Namespace::OuterClass&gt;::encode):
(IPC::ArgumentCoder&lt;Namespace::OuterClass&gt;::decode):
(IPC::ArgumentCoder&lt;Namespace::OtherOuterClass&gt;::encode):
(IPC::ArgumentCoder&lt;Namespace::OtherOuterClass&gt;::decode):
(WTF::isValidOptionSet&lt;EnumNamespace2::OptionSetEnumType&gt;):
(WTF::void&gt;):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
(WebKit::allSerializedEnums):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:

Canonical link: <a href="https://commits.webkit.org/274233@main">https://commits.webkit.org/274233@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90ec5f85a7bb30dfed466a39c155eeec84b77c15

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38288 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17233 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40607 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/40846 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34059 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/19993 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14571 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32301 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/38861 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14548 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12663 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/38194 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12645 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42124 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34851 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34651 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38484 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13225 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/10941 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36682 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14780 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8627 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13635 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14243 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
```diff
@@ -461,6 +461,12 @@ Vector<SerializedTypeInfo> allSerializedTypes()
                 "pendingReads()"_s
             },
         } },
+        { "Namespace::OuterClass"_s, {
```
I feel like this should have ENABLE(OUTER_CONDITION) around it.

aprotyas replied:
For some reason, `[Nested]` class declarations don't reflect this? Changed to a bare class declaration for now, I'm not sure what's special about `[Nested]`.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
```diff
@@ -626,6 +633,7 @@ enum class WebCore_TimingFunction_Subclass : IPC::EncodedVariantIndex {
     , SpringTimingFunction
 };
 
+IGNORE_WARNINGS_BEGIN("missing-noreturn")
```
This is from Andy's change.  Thanks for adding it.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
```diff
@@ -506,7 +512,7 @@ Vector<SerializedEnumInfo> allSerializedEnums()
             static_cast<uint64_t>(EnumNamespace2::OptionSetEnumType::OptionSetSecondValue),
 #endif
-#if !ENABLE(OPTION_SET_SECOND_VALUE)
+#if !(ENABLE(OPTION_SET_SECOND_VALUE))
```
It seems like the parentheses in expression(self) aren't needed.

aprotyas replied:
I made it this way because you could conceive a single condition statement containing multiple logical operands, say if:

```cpp
#if ENABLE(FOO) && ENABLE(BAR)
...
#else
...
#endif
```

We want this to show up as `#if !(ENABLE(FOO) && ENABLE(BAR))` and not `#if !ENABLE(FOO) && ENABLE(BAR)` in our output. Without the parentheses, it's a bit clunky because we'd have to differentiate between simple and compound conditions.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
```diff
@@ -291,3 +290,19 @@ header: "RemoteVideoFrameIdentifier.h"
+    InnerInnerValue,
+#endif
+}
+#endif
```
could you add an else here?  That's worth testing.

aprotyas replied:
Done!

aprotyas committed:
generate-serializers.py should support nested conditions
https://bugs.webkit.org/show_bug.cgi?id=268519
rdar://122056394

Reviewed by Brady Eidson and Alex Christensen.

generate-serializers.py does not currently support nested conditions,
not because of a lack of intention, but just because there hasn't been a
need. However, with the proposed changes for https://webkit.org/b/267428,
the watchOS/tvOS/GTK ports fail to compile because several Apple Pay
enablement conditions are being stacked when declaring data types in
WebCoreArgumentCodersCocoa.serialization.in.

This patch addresses the issue by introducing support for nested
conditions in the serializer generation script. It does so by imbibing
the parse_serialized_types method with the concept of a stack of
conditions, pushing to the stack (or negating the top entry) when an IF
or ELSE is seen, respectively, and only popping from the stack when an
ENDIF is seen. The script now keeps an aggregated condition expression
by walking through the condition stack and logically concatenating
conditions.

* Source/WebKit/Scripts/generate-serializers.py:
(ConditionStackEntry):
(ConditionStackEntry.__init__):
(ConditionStackEntry.expression):
(generate_condition_expression):
(parse_serialized_types):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
(IPC::ArgumentCoder<Namespace::OuterClass>::encode):
(IPC::ArgumentCoder<Namespace::OuterClass>::decode):
(IPC::ArgumentCoder<Namespace::OtherOuterClass>::encode):
(IPC::ArgumentCoder<Namespace::OtherOuterClass>::decode):
(WTF::isValidOptionSet<EnumNamespace2::OptionSetEnumType>):
(WTF::void>):
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
(WebKit::allSerializedEnums):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:

Canonical link: https://commits.webkit.org/274233@main
